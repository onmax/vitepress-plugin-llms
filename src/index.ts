import type { Plugin, ViteDevServer } from 'vite'

import fs from 'node:fs/promises'
import path from 'node:path'

import matter from 'gray-matter'
import { minimatch } from 'minimatch'
import pc from 'picocolors'
import { remark } from 'remark'
import remarkFrontmatter from 'remark-frontmatter'
import { remarkPlease } from './helpers/markdown'

import { remove } from 'unist-util-remove'

import { name as packageName } from '../package.json'

import { millify } from 'millify'
import { approximateTokenSize } from 'tokenx'
import { defaultLLMsTxtTemplate, fullTagRegex, unnecessaryFilesList } from './constants'
import viteDevServerMiddleware from './devserer-middleware'
import {
	generateLLMsFullTxt,
	generateLLMsTxt,
	organizeFilesByDepth,
	generateDirectoryLLMsTxt,
} from './helpers/index'
import log from './helpers/logger'
import { expandTemplate, extractTitle, generateMetadata, getHumanReadableSizeOf } from './helpers/utils'
import type { CustomTemplateVariables, LlmstxtSettings, PreparedFile, VitePressConfig } from './types'

const PLUGIN_NAME = packageName

/**
 * [VitePress](http://vitepress.dev/) plugin for generating raw documentation
 * for **LLMs** in Markdown format which is much lighter and more efficient for LLMs
 *
 * @param [userSettings={}] - Plugin settings.
 *
 * @see https://github.com/okineadev/vitepress-plugin-llms
 * @see https://llmstxt.org/
 */
function llmstxt(userSettings: LlmstxtSettings = {}): [Plugin, Plugin] {
	// Create a settings object with defaults explicitly merged
	const settings: Omit<LlmstxtSettings, 'ignoreFiles' | 'workDir'> & {
		ignoreFiles: string[]
		workDir: string
	} = {
		generateLLMsTxt: true,
		generateLLMsFullTxt: true,
		generateLLMFriendlyDocsForEachPage: true,
		ignoreFiles: [],
		excludeUnnecessaryFiles: true,
		excludeIndexPage: true,
		excludeBlog: true,
		excludeTeam: true,
		workDir: undefined as unknown as string,
		stripHTML: true,
		depth: 1,
		minFilesPerChunk: 2,
		includeNavigation: true,
		...userSettings,
	}

	// Store the resolved Vite config
	let config: VitePressConfig

	// Set to store all markdown file paths
	const mdFiles: Set<string> = new Set()

	// Flag to identify which build we're in
	let isSsrBuild = false

	return [
		{
			enforce: 'pre',
			name: `${PLUGIN_NAME}:llm-tags`,

			/**
			 * Processes each file that Vite transforms and collects markdown files.
			 *
			 * @param content - The file content.
			 * @param id - The file identifier (path).
			 * @returns null if the file is processed, otherwise returns the original content.
			 */
			async transform(content, id) {
				const orig = content

				if (!id.endsWith('.md') || !id.startsWith(settings.workDir)) {
					return null
				}

				if (settings.ignoreFiles?.length) {
					const shouldIgnore = await Promise.all(
						settings.ignoreFiles.map(async (pattern) => {
							if (typeof pattern === 'string') {
								return minimatch(path.relative(settings.workDir, id), pattern)
							}
							return false
						}),
					)

					if (shouldIgnore.some((result) => result === true)) {
						return null
					}
				}

				// Ensure content is a string before processing
				const contentStr = typeof content === 'string' ? content : String(content)

				// strip content between <llm-only> and </llm-only>
				const modifiedContent = contentStr
					.replace(fullTagRegex('llm-only', 'g'), '')
					.replace(fullTagRegex('llm-exclude', 'g'), '$1')
				// remove <llm-exclude> tags, keep the content
				// modifiedContent = modifiedContent

				// Add markdown file path to our collection
				mdFiles.add(id)

				return modifiedContent !== orig ? { code: modifiedContent, map: null } : null
			},
		},
		{
			name: PLUGIN_NAME,
			// Run after all other plugins
			enforce: 'post',

			/** Resolves the Vite configuration and sets up the working directory. */
			configResolved(resolvedConfig) {
				config = resolvedConfig as VitePressConfig
				if (settings.workDir) {
					settings.workDir = path.resolve(config.vitepress.srcDir, settings.workDir)
				} else {
					settings.workDir = config.vitepress.srcDir
				}

				if (settings.excludeUnnecessaryFiles) {
					settings.excludeIndexPage && settings.ignoreFiles.push(...unnecessaryFilesList.indexPage)
					settings.excludeBlog && settings.ignoreFiles.push(...unnecessaryFilesList.blogs)
					settings.excludeTeam && settings.ignoreFiles.push(...unnecessaryFilesList.team)
				}

				// Detect if this is the SSR build
				isSsrBuild = !!resolvedConfig.build?.ssr

				log.info(
					`${pc.bold(PLUGIN_NAME)} initialized ${isSsrBuild ? pc.dim('(SSR build)') : pc.dim('(client build)')} with workDir: ${pc.cyan(settings.workDir)}`,
				)
			},

			/** Configures the development server to handle `llms.txt` and markdown files for LLMs. */
			async configureServer(server: ViteDevServer) {
				server.middlewares.use(viteDevServerMiddleware(config?.vitepress))
				log.info('Dev server configured for serving plain text docs for LLMs')
			},

			/**
			 * Resets the collection of markdown files when the build starts.
			 * This ensures we don't include stale data from previous builds.
			 */
			buildStart() {
				mdFiles.clear()
				log.info('Build started, file collection cleared')
			},

			/**
			 * Runs only in the client build (not SSR) after completion.
			 * This ensures the processing happens exactly once.
			 */
			async generateBundle() {
				// Skip processing during SSR build
				if (isSsrBuild) {
					log.info('Skipping LLMs docs generation in SSR build')
					return
				}

				// resolve the sidebar option before reading `mdFiles`
				// in order to process files from content loaders used in the sidebar function
				const resolvedSidebar =
					settings.sidebar instanceof Function
						? await settings.sidebar(config?.vitepress?.userConfig?.themeConfig?.sidebar)
						: settings.sidebar

				const outDir = config.vitepress?.outDir ?? 'dist'

				// Create output directory if it doesn't exist
				try {
					await fs.access(outDir)
				} catch {
					log.info(`Creating output directory: ${pc.cyan(outDir)}`)
					await fs.mkdir(outDir, { recursive: true })
				}

				const mdFilesList = Array.from(mdFiles)
				const fileCount = mdFilesList.length

				// Skip if no files found
				if (fileCount === 0) {
					log.warn(
						`No markdown files found to process. Check your \`${pc.bold('workDir')}\` and \`${pc.bold('ignoreFiles')}\` settings.`,
					)
					return
				}

				log.info(
					`Processing ${pc.bold(fileCount.toString())} markdown files from ${pc.cyan(settings.workDir)}`,
				)

				const preparedFiles: PreparedFile[] = await Promise.all(
					mdFilesList.map(async (file) => {
						const content = await fs.readFile(file, 'utf-8')

						const markdownProcessor = remark()
							.use(remarkFrontmatter)
							.use(remarkPlease('unwrap', 'llm-only'))
							.use(remarkPlease('remove', 'llm-exclude'))

						if (settings.stripHTML) {
							// Strip HTML tags
							markdownProcessor.use(() => {
								return (tree) => {
									remove(tree, { type: 'html' })
									return tree
								}
							})
						}

						const processedMarkdown = matter(String(await markdownProcessor.process(content)))

						// Extract title from frontmatter or use the first heading
						const title = extractTitle(processedMarkdown)?.trim() || 'Untitled'
						const filePath =
							path.basename(file) === 'index.md' && path.dirname(file) !== settings.workDir
								? `${path.dirname(file)}.md`
								: file

						return { path: filePath, title, file: processedMarkdown }
					}),
				)

				// Sort files by title for better organization
				preparedFiles.sort((a, b) => a.title.localeCompare(b.title))

				const tasks: Promise<void>[] = []

				if (settings.generateLLMsTxt) {
					// Organize files by depth for hierarchical llms.txt generation
					const directoryChunks = organizeFilesByDepth(
						preparedFiles,
						settings.workDir,
						settings.depth || 1,
						settings.minFilesPerChunk || 2,
					)

					// Generate llms.txt files for each directory chunk
					for (const chunk of directoryChunks) {
						const isRoot = chunk.dirPath === ''
						const llmsTxtPath = isRoot
							? path.resolve(outDir, 'llms.txt')
							: path.resolve(outDir, chunk.dirPath, 'llms.txt')

						const templateVariables: CustomTemplateVariables = {
							title: settings.title,
							description: settings.description,
							details: settings.details,
							toc: settings.toc,
							...settings.customTemplateVariables,
						}

						tasks.push(
							(async () => {
								const displayPath = isRoot ? 'llms.txt' : `${chunk.dirPath}/llms.txt`
								log.info(`Generating ${pc.cyan(displayPath)}...`)

								// Create directory if it doesn't exist
								if (!isRoot) {
									await fs.mkdir(path.dirname(llmsTxtPath), { recursive: true })
								}

								const llmsTxt = await generateDirectoryLLMsTxt(chunk, {
									indexMd: path.resolve(settings.workDir, 'index.md'),
									srcDir: settings.workDir,
									LLMsTxtTemplate: settings.customLLMsTxtTemplate || defaultLLMsTxtTemplate,
									templateVariables,
									vitepressConfig: config?.vitepress?.userConfig,
									domain: settings.domain,
									sidebar: resolvedSidebar,
									linksExtension: !settings.generateLLMFriendlyDocsForEachPage ? '.html' : undefined,
									cleanUrls: config.cleanUrls,
									includeNavigation: settings.includeNavigation,
								})

								await fs.writeFile(llmsTxtPath, llmsTxt, 'utf-8')

								log.success(
									expandTemplate(
										'Generated {file} (~{tokens} tokens, {size}) with {fileCount} documentation links',
										{
											file: pc.cyan(displayPath),
											tokens: pc.bold(millify(approximateTokenSize(llmsTxt))),
											size: pc.bold(getHumanReadableSizeOf(llmsTxt)),
											fileCount: pc.bold(chunk.files.length.toString()),
										},
									),
								)
							})(),
						)
					}
				}

				// Generate llms-full.txt - all content in one file
				if (settings.generateLLMsFullTxt) {
					const llmsFullTxtPath = path.resolve(outDir, 'llms-full.txt')

					tasks.push(
						(async () => {
							log.info(`Generating full documentation bundle (${pc.cyan('llms-full.txt')})...`)

							const llmsFullTxt = await generateLLMsFullTxt(preparedFiles, {
								srcDir: settings.workDir,
								domain: settings.domain,
								linksExtension: !settings.generateLLMFriendlyDocsForEachPage ? '.html' : undefined,
								cleanUrls: config.cleanUrls,
							})

							// Write content to llms-full.txt
							await fs.writeFile(llmsFullTxtPath, llmsFullTxt, 'utf-8')

							log.success(
								expandTemplate(
									'Generated {file} (~{tokens} tokens, {size}) with {fileCount} markdown files',
									{
										file: pc.cyan('llms-full.txt'),
										tokens: pc.bold(millify(approximateTokenSize(llmsFullTxt))),
										size: pc.bold(getHumanReadableSizeOf(llmsFullTxt)),
										fileCount: pc.bold(fileCount.toString()),
									},
								),
							)
						})(),
					)
				}

				if (settings.generateLLMFriendlyDocsForEachPage) {
					tasks.push(
						...preparedFiles.map(async (file) => {
							const relativePath = path.relative(settings.workDir, file.path)
							try {
								const mdFile = file.file
								const targetPath = path.resolve(outDir, relativePath)

								await fs.mkdir(path.dirname(targetPath), { recursive: true })

								await fs.writeFile(
									targetPath,
									matter.stringify(
										mdFile.content,
										await generateMetadata(mdFile, {
											domain: settings.domain,
											filePath: relativePath,
											linksExtension: '.md',
											cleanUrls: config.cleanUrls,
										}),
									),
								)

								log.success(`Processed ${pc.cyan(relativePath)}`)
							} catch (error) {
								log.error(`Failed to process ${pc.cyan(relativePath)}: ${(error as Error).message}`)
							}
						}),
					)
				}

				if (tasks.length) {
					await Promise.all(tasks)
				}
			},
		},
	]
}

export default llmstxt
