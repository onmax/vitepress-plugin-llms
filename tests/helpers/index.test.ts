import { describe, expect, it, mock } from 'bun:test'

import { defaultLLMsTxtTemplate } from '../../src/constants'

import { mockedFs } from '../mocks/fs'

mockedFs.default.readFile.mockReturnValue(Promise.resolve(fakeIndexMd))

mock.module('node:fs/promises', () => mockedFs)

import {
	generateLLMsFullTxt,
	generateLLMsTxt,
	organizeFilesByDepth,
	generateNavigationSection,
	generateDirectoryLLMsTxt,
	// @ts-ignore
} from '../../src/helpers'
import {
	fakeCustomLlmsTxtTemplate,
	fakeIndexMd,
	preparedFilesSample,
	preparedFilesDepthSample,
	sampleDomain,
	srcDir,
} from '../resources'

describe('generateLLMsTxt', () => {
	it('generates a `llms.txt` file', async () => {
		expect(
			await generateLLMsTxt(preparedFilesSample({ srcDir }).slice(1), {
				indexMd: `${srcDir}/index.md`,
				srcDir,
				LLMsTxtTemplate: defaultLLMsTxtTemplate,
				templateVariables: {},
				vitepressConfig: {},
			}),
		).toMatchSnapshot()
	})
	it('works correctly with a custom template', async () => {
		expect(
			await generateLLMsTxt(preparedFilesSample({ srcDir }).slice(1), {
				indexMd: `${srcDir}/index.md`,
				srcDir,
				LLMsTxtTemplate: fakeCustomLlmsTxtTemplate,
				templateVariables: {},
				vitepressConfig: {},
			}),
		).toMatchSnapshot()
	})
	it('works correctly with a custom template variables', async () => {
		expect(
			await generateLLMsTxt(preparedFilesSample({ srcDir }), {
				indexMd: `${srcDir}/index.md`,
				srcDir,
				LLMsTxtTemplate: defaultLLMsTxtTemplate,
				templateVariables: { title: 'foo', description: 'bar', toc: 'zoo' },
				vitepressConfig: {},
			}),
		).toMatchSnapshot()
	})

	it('works correctly with a custom template and variables', async () => {
		expect(
			await generateLLMsTxt(preparedFilesSample({ srcDir }), {
				indexMd: `${srcDir}/index.md`,
				srcDir,
				LLMsTxtTemplate: '# {foo}\n\n**{bar}**\n\n{zoo}',
				templateVariables: { title: 'foo', description: 'bar', toc: 'zoo' },
				vitepressConfig: {},
			}),
		).toMatchSnapshot()
	})
})

describe('generateLLMsFullTxt', () => {
	it('generates a `llms-full.txt` file', async () => {
		expect(
			await generateLLMsFullTxt(preparedFilesSample({ srcDir }).slice(1), {
				srcDir,
			}),
		).toMatchSnapshot()
	})

	it('correctly attaches the domain to URLs in context', async () => {
		expect(
			await generateLLMsFullTxt(preparedFilesSample({ srcDir }).slice(1), {
				srcDir,
				domain: sampleDomain,
			}),
		).toMatchSnapshot()
	})
})

describe('organizeFilesByDepth', () => {
	it('organizes files by depth level 1 (root only)', () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 1, 2)

		expect(chunks).toHaveLength(1)
		expect(chunks[0].dirPath).toBe('')
		expect(chunks[0].depth).toBe(1)
		expect(chunks[0].files).toHaveLength(files.length)
	})

	it('organizes files by depth level 2', () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 2, 2)

		expect(chunks.length).toBeGreaterThan(1)

		// Should have root chunk
		const rootChunk = chunks.find((c) => c.dirPath === '')
		expect(rootChunk).toBeDefined()

		// Should have guide, api, examples chunks
		const guideChunk = chunks.find((c) => c.dirPath === 'guide')
		const apiChunk = chunks.find((c) => c.dirPath === 'api')
		const examplesChunk = chunks.find((c) => c.dirPath === 'examples')

		expect(guideChunk).toBeDefined()
		expect(apiChunk).toBeDefined()
		expect(examplesChunk).toBeDefined()
	})

	it('organizes files by depth level 3', () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 3, 2)

		// Should have more chunks including nested directories
		const advancedChunk = chunks.find((c) => c.dirPath === 'guide/advanced')
		const v2Chunk = chunks.find((c) => c.dirPath === 'api/v2')

		expect(advancedChunk).toBeDefined()
		expect(v2Chunk).toBeDefined()
	})

	it('respects minimum files per chunk', () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 3, 5)

		// With minFilesPerChunk = 5, some smaller directories should be filtered out
		const nonRootChunks = chunks.filter((c) => c.dirPath !== '')
		expect(nonRootChunks.length).toBeLessThan(chunks.length)
	})

	it('sets correct navigation context', () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 3, 2)

		const guideChunk = chunks.find((c) => c.dirPath === 'guide')
		const advancedChunk = chunks.find((c) => c.dirPath === 'guide/advanced')
		const apiChunk = chunks.find((c) => c.dirPath === 'api')
		const v2Chunk = chunks.find((c) => c.dirPath === 'api/v2')

		// Verify guide chunk (depth 2)
		expect(guideChunk?.parentPath).toBe('')
		expect(guideChunk?.siblingPaths).toContain('api')
		expect(guideChunk?.siblingPaths).toContain('examples')
		expect(guideChunk?.childPaths).toContain('guide/advanced')

		// Verify advanced chunk (depth 3, child of guide)
		expect(advancedChunk?.parentPath).toBe('guide')
		// Advanced chunk has no siblings at depth 3 with parent 'guide'
		expect(advancedChunk?.siblingPaths).toEqual([])

		// Verify api chunk (depth 2)
		expect(apiChunk?.siblingPaths).toContain('guide')
		expect(apiChunk?.siblingPaths).toContain('examples')
		expect(apiChunk?.childPaths).toContain('api/v2')

		// Verify v2 chunk (depth 3, child of api)
		expect(v2Chunk?.parentPath).toBe('api')
		// V2 chunk has no siblings at depth 3 with parent 'api'
		expect(v2Chunk?.siblingPaths).toEqual([])
	})
})

describe('generateNavigationSection', () => {
	it('generates navigation links for parent, siblings, and children', () => {
		const navigation = generateNavigationSection({
			currentDirPath: 'guide',
			parentPath: '',
			siblingPaths: ['api', 'examples'],
			childPaths: ['guide/advanced'],
			domain: undefined,
			srcDir,
			cleanUrls: false,
		})

		expect(navigation).toContain('Documentation Overview')
		expect(navigation).toContain('/llms.txt')
		expect(navigation).toContain('Api')
		expect(navigation).toContain('/api/llms.txt')
		expect(navigation).toContain('Examples')
		expect(navigation).toContain('/examples/llms.txt')
		expect(navigation).toContain('Advanced')
		expect(navigation).toContain('/guide/advanced/llms.txt')
	})

	it('generates navigation links with domain', () => {
		const navigation = generateNavigationSection({
			currentDirPath: 'api',
			parentPath: '',
			siblingPaths: ['guide'],
			childPaths: ['api/v2'],
			domain: sampleDomain,
			srcDir,
			cleanUrls: false,
		})

		expect(navigation).toContain(`${sampleDomain}/llms.txt`)
		expect(navigation).toContain(`${sampleDomain}/guide/llms.txt`)
		expect(navigation).toContain(`${sampleDomain}/api/v2/llms.txt`)
	})

	it('returns empty string when no navigation links exist', () => {
		const navigation = generateNavigationSection({
			currentDirPath: '',
			parentPath: '',
			siblingPaths: [],
			childPaths: [],
			domain: undefined,
			srcDir,
			cleanUrls: false,
		})

		expect(navigation).toBe('')
	})
})

describe('generateDirectoryLLMsTxt', () => {
	it('generates llms.txt for root directory', async () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 1, 2)
		const rootChunk = chunks.find((c) => c.dirPath === '')!

		const result = await generateDirectoryLLMsTxt(rootChunk, {
			indexMd: `${srcDir}/index.md`,
			srcDir,
			templateVariables: { title: 'Test Documentation' },
			includeNavigation: true,
		})

		expect(result).toContain('# Test Documentation')
		expect(result).toContain('## Table of Contents')
		expect(result).toMatchSnapshot()
	})

	it('generates llms.txt for subdirectory with navigation', async () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 2, 2)
		const guideChunk = chunks.find((c) => c.dirPath === 'guide')!

		const result = await generateDirectoryLLMsTxt(guideChunk, {
			indexMd: `${srcDir}/index.md`,
			srcDir,
			templateVariables: {},
			includeNavigation: true,
		})

		expect(result).toContain('# Guide Documentation')
		expect(result).toContain('## Navigation')
		expect(result).toContain('Documentation Overview')
		expect(result).toContain('## Documentation')
		expect(result).toMatchSnapshot()
	})

	it('generates llms.txt without navigation when disabled', async () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 2, 2)
		const apiChunk = chunks.find((c) => c.dirPath === 'api')!

		const result = await generateDirectoryLLMsTxt(apiChunk, {
			indexMd: `${srcDir}/index.md`,
			srcDir,
			templateVariables: {},
			includeNavigation: false,
		})

		expect(result).toContain('# Api Documentation')
		expect(result).not.toContain('## Navigation')
		expect(result).toContain('## Documentation')
		expect(result).toMatchSnapshot()
	})

	it('generates llms.txt with domain in navigation links', async () => {
		const files = preparedFilesDepthSample({ srcDir })
		const chunks = organizeFilesByDepth(files, srcDir, 3, 2)
		const advancedChunk = chunks.find((c) => c.dirPath === 'guide/advanced')!

		const result = await generateDirectoryLLMsTxt(advancedChunk, {
			indexMd: `${srcDir}/index.md`,
			srcDir,
			domain: sampleDomain,
			templateVariables: {},
			includeNavigation: true,
		})

		expect(result).toContain(sampleDomain)
		expect(result).toContain(`${sampleDomain}/guide/llms.txt`)
		expect(result).toMatchSnapshot()
	})
})
