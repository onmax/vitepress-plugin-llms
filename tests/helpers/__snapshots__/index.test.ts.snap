// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateLLMsTxt generates a \`llms.txt\` file 1`] = `
"# Some cool tool

> Blazing fast frontend tool

## Table of Contents

- [Getting started](/test/getting-started.md): Instructions on how to get started with the tool
- [Quickstart](/test/quickstart.md): Instructions for quick project initialization
- [Some other section](/test/other.md)
"
`;

exports[`generateLLMsTxt works correctly with a custom template 1`] = `
"# Custom title

> Custom description

## TOC

- [Getting started](/test/getting-started.md): Instructions on how to get started with the tool
- [Quickstart](/test/quickstart.md): Instructions for quick project initialization
- [Some other section](/test/other.md)
"
`;

exports[`generateLLMsTxt works correctly with a custom template variables 1`] = `
"# foo

> bar

## Table of Contents

zoo"
`;

exports[`generateLLMsTxt works correctly with a custom template and variables 1`] = `
"# {foo}

**{bar}**

{zoo}"
`;

exports[`generateLLMsFullTxt generates a \`llms-full.txt\` file 1`] = `
"---
url: /test/getting-started.md
description: Instructions on how to get started with the tool
---

# Installation

blah blah blah...

---

---
url: /test/quickstart.md
description: Instructions for quick project initialization
---

# Project initialization

blah blah blah...

---

---
url: /test/other.md
---
# Some cool stuff
"
`;

exports[`generateLLMsFullTxt correctly attaches the domain to URLs in context 1`] = `
"---
url: 'https://example.com/test/getting-started.md'
description: Instructions on how to get started with the tool
---

# Installation

blah blah blah...

---

---
url: 'https://example.com/test/quickstart.md'
description: Instructions for quick project initialization
---

# Project initialization

blah blah blah...

---

---
url: 'https://example.com/test/other.md'
---
# Some cool stuff
"
`;

exports[`generateDirectoryLLMsTxt generates llms.txt for root directory 1`] = `
"# Test Documentation

> Blazing fast frontend tool

## Table of Contents

- [Documentation Home](/index.md): Blazing fast frontend tool
- [Getting started](/guide/getting-started.md): Instructions on how to get started with the tool
- [Installation Guide](/guide/installation.md): Instructions for quick project initialization
- [Advanced Configuration](/guide/advanced/configuration.md)
- [Performance Tuning](/guide/advanced/performance.md)
- [API Reference](/api/reference.md)
- [API Examples](/api/examples.md)
- [Components API](/api/v2/components.md)
- [Hooks API](/api/v2/hooks.md)
- [Basic Example](/examples/basic.md)
- [Advanced Example](/examples/advanced.md)
"
`;

exports[`generateDirectoryLLMsTxt generates llms.txt for subdirectory with navigation 1`] = `
"# Guide Documentation

> Documentation for the guide section

Documentation and guides for guide.

## Navigation

- [Documentation Overview](/llms.txt): Parent documentation section
- [Api](/api/llms.txt): Api documentation
- [Examples](/examples/llms.txt): Examples documentation


## Documentation

- [Getting started](/guide/getting-started.md): Instructions on how to get started with the tool
- [Installation Guide](/guide/installation.md): Instructions for quick project initialization
- [Advanced Configuration](/guide/advanced/configuration.md)
- [Performance Tuning](/guide/advanced/performance.md)
"
`;

exports[`generateDirectoryLLMsTxt generates llms.txt without navigation when disabled 1`] = `
"# Api Documentation

> Documentation for the api section

Documentation and guides for api.
## Documentation

- [API Reference](/api/reference.md)
- [API Examples](/api/examples.md)
- [Components API](/api/v2/components.md)
- [Hooks API](/api/v2/hooks.md)
"
`;

exports[`generateDirectoryLLMsTxt generates llms.txt with domain in navigation links 1`] = `
"# Advanced Documentation

> Documentation for the advanced section

Documentation and guides for advanced.

## Navigation

- [Guide](https://example.com/guide/llms.txt): Parent documentation section


## Documentation

- [Advanced Configuration](https://example.com/guide/advanced/configuration.md)
- [Performance Tuning](https://example.com/guide/advanced/performance.md)
"
`;
